ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+* Rust bridge uses [rb-sys](https://github.com/oxidize-rb/rb-sys) and [magnus](https://github.com/matsadler/magnus)
+  * 💭 Why not just `ruby.h` directly or just `rb-sys`? There are many things difficult to get right with data
+    conversion and we don't feel as if using either library is compromising  our ability to move later if needed.
+* Obvious general repo/project rules apply.
+  * Linting and auto-formatting (Rubocop check/autocorrect but have to customize Rubocop a lot because its defaults are
+    very dogmatic).
+  * Type checking (Steep).
+  * Deployed to RubyGems as separate gems for pure source, x64 linux, x64 darwin, x64 windows (i.e. mingw-ucrt), aarch64
+    linux (i.e. arm64), and arm64 darwin.
+  * CI against supported platforms on oldest and newest Ruby versions supported.
+  * MIT licensed.
+  * Full docs as README.
+  * No need for ruby.temporal.io, https://www.rubydoc.info/gems/temporalio is acceptable.
+  * Full integration test suite.
+
+### Basic Features
+
+By the very first version, the Ruby SDK should have feature parity with every other SDK (except in cases where we have
+decided other features in other SDKs will never be ported). Here's a non-exhaustive list of some features that includes
+that are not discussed later:
+
+* Interceptors and an OpenTelemetry implementation
+* Custom metrics and optional metrics buffer for custom processing
+* Typed search attributes, memos, headers, etc
+* API key, RPC metadata, HTTP CONNECT proxy support, etc
+* Loggers with contextual information
+* Full error hierarchy with `Temporalio::Error` as the root
+* ❓ Are there any Ruby libraries so widely adopted they might as well be in the standard library and therefore deserve
+  optional support (e.g. like Go/Python SDK contrib or .NET SDK extensions)?
+
+## Runtime
+
+* `Temporalio::Runtime` is the full Core runtime.
+* Will have class instance variable of `default` that can be set and is lazily created upon first access.
+  * Client (technically connection) accepts a runtime and if unset it defaults to `default`
+* Initialization of the runtime class starts a Ruby thread that just runs the command loop from Rust.
+  * 💭 Why does a runtime need to eat a Ruby thread? The Rust side needs to have a way to make Ruby calls, and only Ruby
+    can start a Ruby thread. So on the Rust side there is essentially a loop with a channel waiting without holding the
+    GVL with things to run on the Ruby side.
+  * 💭 Why do this in a constructor? That is the Ruby way, a constructed object is ready for use.
+  * 💭 Why only a single thread? We expect callbacks to Ruby from Rust to be short (hopefully only a block invocation
+    that is a `Queue.push`) so we don't expect making all of these run serial is an issue. However if it does become an
+    issue, we can use multiple threads and a Rust approach which allows multiple channel consumers.
+* Users can create their own runtime with their own telemetry options and such and pass that to clients.
```

Comment:
I've never seen the need to create more than a single runtime in our SDKs. Seems like we can avoid this extra API surface to expose accepting a runtime and just have all constructs reference a singleton.

Reply:
They will reference a lazy global which is basically a singleton. This is deemed acceptable in other core-based SDKs and so should be here too, no need to make Ruby special.

Reply:
Yeah I'm fine with it being a singleton/lazy global under the hood. I just don't want us to design ourselves into a corner.